packages/@lwc/engine/dom/src/apis/build-custom-element-constructor.ts

```diff
@@ -38,3 +49,56 @@ export function deprecatedBuildCustomElementConstructor(
 
     return Ctor.CustomElementConstructor;
 }
+
+export function buildCustomElementConstructor(Ctor: ComponentConstructor): HTMLElementConstructor {
+    const def = getComponentInternalDef(Ctor);
+
+    // generating the hash table for attributes to avoid duplicate fields and facilitate validation
+    // and false positives in case of inheritance.
+    const attributeToPropMap: Record<string, string> = create(null);
+    for (const propName in def.props) {
+        attributeToPropMap[getAttrNameFromPropName(propName)] = propName;
+    }
+
+    return class extends def.bridge {
+        constructor() {
+            super();
+            createVM(this, def, {
+                mode: 'open',
+                isRoot: true,
+                owner: null,
+                tagName: this.tagName,
```

Comment:
where is tagName defined?

Reply:
We're extending `Element` and it's defined there.